### ISSUE TITLE
no (or more than one) schemas match

### ISSUE THREAD

**Issue by shachr** [mafintosh/is-my-json-valid#82]
Given the following schema:
```javascript
{
        id:"/api/accounts/register",
        additionalProperties: false,
        "oneOf": ["#/definitions/social","#/definitions/site"],

        definitions: {
            social: {
                additionalProperties: false,
                required: true,
                type: 'object',
                properties: {
                    provider: { enum: [ "facebook" ], required:true }
                }
            },
            site: {
                additionalProperties: false,
                required: true,
                type: 'object',
                properties: {
                    username: { type: "string", required:true },
                    password: { type: "string", required:true },

                    firstname: { type: "string" },
                    lastname: { type: "string" },
                    email: { type: "string" },
                    isSubscribed: { type: "boolean" }
                }
            }
        }

    }
```

i get the following validation error:
```javascript
[
      {
        "field": "data",
        "message": "no (or more than one) schemas match",
        "value": {
          
        }
      }
    ]
```

i would expect to get the following or something similar which is much more useful:
```javascript
[
      {
        "field": "data",
        "message": "no (or more than one) schemas match",
        "value": {
          
        }
      },

     {
        "field": "data",
        "type": "oneOf",
        "message": "one of",
        "value": [
             { "message": "is required", "field": ["username","password"]  },
             { "message": "is required", "field": ["provider"]  },
        ]
     }
    ]
```

this will allow me to show the user that either ( username and password ) or (provider) are missing.

**Comment by mattanbi** [comment]
Hi, does anyone know the status of this issue? I get the same response.

**Comment by mafintosh** [comment]
i agree that this should be better. this was the easiest way to implement it when i originally did it. if anyone wants to do a pr to fix this let me know

**Comment by mattanbi** [comment]
I'll be happy to. Can you shortly explain the validate and error methods so I understand what should be changed there?
  if (node.oneOf && node.oneOf.length) {
                var prev = gensym('prev')
                var passes = gensym('passes')

                validate
                ('var %s = errors', prev)
                ('var %s = 0', passes)

                node.oneOf.forEach(function(sch, i) {
                    visit(name, sch, false, false)
                    validate('if (%s === errors) {', prev)
                    ('%s++', passes)
                    ('} else {')
                    ('errors = %s', prev)
                    ('}')
                })

                validate('if (%s !== 1) {', passes)
                error('no (or more than one) schemas match')
                validate('}')
            }

**Comment by mafintosh** [comment]
@mattanbi basically what that snippet does it that goes through all the schemas defined by the `oneOf` and checks if that schema matches. If more and 1 or 0 schemas match it'll produce an error by calling the `error` function. it ignores the actual error messages produced by the subschemas since as long as a single schema from the oneOf matches everything is fine

**Comment by mattanbi** [comment]
Good, thanks. Can you point out how I get the errors (that are currently ignored, as you say)? Because when I debug the validate method doesn't return anything.
I don't fully understand where it is actually determent that there is an error.

**Comment by simranjeet2589** [comment]
I am also facing same issue and tried something to fix it. Here is the pr
https://github.com/mafintosh/is-my-json-valid/pull/91
